src/librustc_mir/const_eval.rs

```diff
@@ -336,6 +336,20 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
         _unwind: Option<mir::BasicBlock> // unwinding is not supported in consts
     ) -> InterpResult<'tcx, Option<&'mir mir::Body<'tcx>>> {
         debug!("eval_fn_call: {:?}", instance);
+
+        // If this function is a `const fn` then as an optimization we can query this
+        // evaluation immediately.
+        //
+        // For the moment we only do this for functions which take no arguments
+        // (or all arguments are ZSTs) so that we don't memoize too much.
```

Comment:
I'm guessing the other thing you check for, longer-term, is that there are no relocations in the arguments, so no memory outside of the call could be modified from inside.

Reply:
I don't think we want to do that. Otherwise every single const fn call without relocations will get monomorphized, likely causing a significant increase in memory usage (e.g. if you call `i32::abs` in a loop over 0..10000, you'll get a cached query for every single call).

While the same thing is true for different ZSTs being passed to generic functions, I think that's less extreme since the user needs to write down each type anyway